### PR TITLE
feat: singleRecordQuery

### PR DESCRIPTION
**Pull request by mshanemc**
Add a function to Connection.  @W-8529337@

Several plugins use conn.query get results back and optimistically use records[0].  This reduces duplicative checks and messaging for 0 and >1 records.

### CODE REVIEW

File: src/connection.ts [hunk 1]
```diff
@@ -297,6 +297,46 @@ export class Connection extends JSForceConnection {
...
+   */
+  public async singleRecordQuery<T>(
+    soql: string,
+    options: SingleRecordQueryOptions = {
+      choiceField: 'Name',
+    }
+  ): Promise<T> {
+    const result = options.tooling ? await this.tooling.query<T>(soql) : await this.query<T>(soql);
+    if (result.totalSize === 0) {
+      throw new SfdxError(`No records found for ${soql}`, SingleRecordQueryErrors.NoRecords);
```
Comment (peternhale) [inline src/connection.ts:316]: Given this function is for a single record, consider `No record found for ...`

File: src/connection.ts [hunk 2]
```diff
@@ -297,6 +297,46 @@ export class Connection extends JSForceConnection {
...
+    }
+    if (result.totalSize > 1) {
+      throw new SfdxError(
+        options.returnChoicesOnMultiple && options.choiceField
+          ? `Multiple records found. ${result.records
+              .map((item) => item[(options.choiceField ?? 'Name') as keyof T])
```
Comment (peternhale) [inline src/connection.ts:322]: Given condition of `options.returnChoicesOnMultiple && options.choiceField` it is not possible for `options.choiceField` to be falsey here.

Comment (mshanemc) [reply src/connection.ts]: yeah, this was from before I put a default on that.  fixing both

File: src/connection.ts [hunk 3]
```diff
@@ -297,6 +297,44 @@ export class Connection extends JSForceConnection {
...
+}
+
+export const SingleRecordQueryErrors = {
+  NoRecords: 'SingleRecordQuery_NoRecords',
+  MultipleRecords: 'SingleRecordQuery_MultipleRecords',
+};
+export interface SingleRecordQueryOptions {
+  tooling?: boolean;
+  returnChoicesOnMultiple?: boolean;
+  choiceField?: string; // defaults to Name
```
Comment (peternhale) [inline src/connection.ts:337]: Where is default value applied?

Comment (mshanemc) [reply src/connection.ts:337]: line 311

File: test/unit/connectionTest.ts [hunk 4]
```diff
@@ -206,4 +206,74 @@ describe('Connection', () => {
...
+    }
+  });
+
+  it('singleRecordQuery handles tooling api flag', async () => {
+    const mockSingleRecord = {
+      id: '123',
+      name: 'testName',
+    };
+    requestMock.returns(Promise.resolve({ totalSize: 1, records: [mockSingleRecord] }));
+    // const querySpy = $$.SANDBOX.spy(jsforce.Connection.prototype, 'query');
```
Comment (shetzel) [inline test/unit/connectionTest.ts:268]: remove this commented code?

**Review by shetzel — APPROVED** [review]
approved pending comment resolution

**Review by shetzel — APPROVED** [review]

**Pull request merged**